aesara/tensor/nlinalg.py

```diff
@@ -823,6 +824,69 @@ def tensorsolve(a, b, axes=None):
     return TensorSolve(axes)(a, b)
 
 
+class MatMulOp(Op):
+    __props__ = ("dtype",)
+
+    def __init__(self, dtype=None):
+        self.dtype = dtype
+
+    def make_node(self, a, b):
+        a = at.as_tensor_variable(a)
+        b = at.as_tensor_variable(b)
+
+        if 0 in (a.ndim, b.ndim):
+            raise ValueError("inputs to `matmul` cannot be scalar.")
+
+        if 2 == a.ndim == b.ndim:
+            out = matrix(dtype=self.dtype)
+        elif a.ndim > 2 or b.ndim > 2:
+            out_shape = broadcast_shape(a, b)
+            out = at.TensorType(
+                dtype=self.dtype,
+                broadcastable=(False,) * len(out_shape),
+            )()
```

Comment:
`broadcast_shape` is a relatively expensive thing to do just to get the number of output dimensions.

Reply:
You're right, i didn't think that one through. Resolved in latest commit.